epoch.go

```diff
@@ -1696,12 +1804,108 @@ func (e *Epoch) storeProposal(block Block) bool {
+	e.maybeLoadFutureMessages()
+
 	return true
 }
 
+// HandleRequest processes a request and returns a response. It also sends a response to the sender.
+func (e *Epoch) HandleReplicationRequest(req *ReplicationRequest, from NodeID) *ReplicationResponse {
+	if !e.ReplicationEnabled {
+		return &ReplicationResponse{}
+	}
+	response := &ReplicationResponse{}
+	if req.FinalizationCertificateRequest != nil {
+		response.FinalizationCertificateResponse = e.handleFinalizationCertificateRequest(req.FinalizationCertificateRequest)
+	}
+
+	msg := &Message{ReplicationResponse: response}
+	e.Comm.SendMessage(msg, from)
+	return response
+}
+
+func (e *Epoch) handleFinalizationCertificateRequest(req *FinalizationCertificateRequest) *FinalizationCertificateResponse {
+	e.Logger.Debug("Received finalization certificate request", zap.Int("num seqs", len(req.Sequences)))
+	seqs := req.Sequences
+	slices.Sort(seqs)
+
+	data := make([]FinalizedBlock, len(seqs))
+	for i, seq := range seqs {
+		block, fCert, found := e.locateSequence(seq)
+		if !found {
+			// since we are sorted, we can break early
+			data = data[:i]
+			break
+		}
+		data[i] = FinalizedBlock{
+			Block: block,
+			FCert: fCert,
+		}
+	}
+
+	return &FinalizationCertificateResponse{
+		Data: data,
+	}
+}
+
+func (e *Epoch) locateSequence(seq uint64) (Block, FinalizationCertificate, bool) {
+	for _, round := range e.rounds {
+		blockSeq := round.block.BlockHeader().Seq
+		if blockSeq == seq {
+			if round.fCert == nil {
+				break
+			}
+
+			return round.block, *round.fCert, true
+		}
+	}
+
+	block, fCert, exists := e.Storage.Retrieve(seq)
+	return block, fCert, exists
+}
+
+func (e *Epoch) handleReplicationResponse(resp *ReplicationResponse, from NodeID) error {
+	var err error
+	if resp.FinalizationCertificateResponse != nil {
+		err = e.handleFinalizationCertificateResponse(resp.FinalizationCertificateResponse, from)
+	}
+	return err
+}
+
+func (e *Epoch) handleFinalizationCertificateResponse(resp *FinalizationCertificateResponse, from NodeID) error {
+	e.Logger.Debug("Received finalization certificate response", zap.String("from", from.String()), zap.Int("num seqs", len(resp.Data)))
+	for _, data := range resp.Data {
+		if e.round+e.maxRoundWindow < data.FCert.Finalization.Seq {
+			e.Logger.Debug("Received finalization certificate for a round that is too far ahead", zap.Uint64("seq", data.FCert.Finalization.Seq))
+			// we are too far behind, we should ignore this message
+			continue
+		}
+
+		err := e.replicationState.StoreFinalizedBlock(data)
+		if err != nil {
+			e.Logger.Info("Failed to store sequence data", zap.Error(err), zap.Uint64("seq", data.FCert.Finalization.Seq), zap.String("from", from.String()))
+			continue
+		}
+	}
+
+	e.processReplicationState()
+	return nil
+}
+
+func (e *Epoch) processReplicationState() {
+	nextSeqToCommit := e.Storage.Height()
+	finalizedBlock, ok := e.replicationState.receivedFinalizationCertificates[nextSeqToCommit]
+	if !ok {
+		return
+	}
+
```

Comment:
isn't this where we probably should delete `finalizedBlock` from `receivedFinalizationCertificates` ? I don't see where we delete it.